### ISSUE TITLE
Abs function on cpp_bin_float throwing C2665 error

### ISSUE THREAD

**Issue by gpive** [boostorg/multiprecision#423]
Build is done using MSVC 2015.

This code throws a C2665 error on Boost 1.78.0 but worked on Boost 1.70.0.

typedef boost::multiprecision::number<boost::multiprecision::backends::cpp_bin_float<256, boost::multiprecision::digit_base_2>> high_prec_double;
high_prec_double x(0.0);
abs(x);

**Comment by jzmaddock** [comment]
Works for me with:

Microsoft (R) C/C++ Optimizing Compiler Version 19.00.24245 for x64

**Comment by jzmaddock** [comment]
Ping?  Can the OP provide a test case please?

**Comment by ckormanyos** [comment]
That's an older setup.

To me it looks like the compiler can't resolve the exact instantiation of `abs`to use.

Could you maybe try to place the line `using std::abs;` as the first line within the scope of the subroutine `TestAbs()`?

**Comment by gpive** [comment]
> That's an older setup.
> 
> To me it looks like the compiler can't resolve the exact instantiation of `abs`to use.
> 
> Could you maybe try to place the line `using std::abs;` as the first line within the scope of the subroutine `TestAbs()`?

Explicitly defining the call as std::abs also failed with the same error.

I'm removing these templates now but not sure why this particular one is causing issues.  Was there to save me having to define operator!=, operator>, operator>=, and operator <= when I already defined operator== and operator<.

**Comment by ckormanyos** [comment]
I tried 1.70 on a newer MSVC compiler and had no problem.

Out of historical interest, it's kind of interesting.

I had meant try a few things such as:

```
void TestAbs()
{
  using std::abs;

  high_prec_double x(0.0);

  abs(x);
}
```

I think this might, however, not make a difference i nyour particular use case. I'm pretty sure Multiprecision's namespace version of `abs` is/was present at the time. I am not exactly sure why the compiler can't resolve it.

**Comment by gpive** [comment]
Whatever is causing the problem, I at least have a work around.

In the above comment, I tried "std::abs(x)" instead of the using function but it'd be the same result.

Also moving from that nuclear set of templates to a more conditional set of macros should help.  Just interesting this is the only function I've ever found it to cause a problem on.

#define AT5_OP_NE(x) bool operator!=(const x& rhs) const { return !(operator==(rhs));}
#define AT5_OP_LTE(x) bool operator<=(const x& rhs) const { return !(rhs.operator<(*this));}
#define AT5_OP_GT(x) bool operator>(const x& rhs) const { return (rhs.operator<(*this));}
#define AT5_OP_GTE(x) bool operator>=(const x& rhs) const { return !(operator<(rhs));}
#define AT5_OP_COMP(x) AT5_OP_NE(x) AT5_OP_LTE(x) AT5_OP_GT(x) AT5_OP_GTE(x)

**Comment by ckormanyos** [comment]
>  have a work around

That is best. There is, in fact, not usually a strong effort to post-patch a 2 year old Boost. Unfortunately, we can't at the moment seem to find an equitable fix, but please keep going with the workaround.

One last query from my side, wondering if using `fabs` might be a viable option and if the compiler resolution is better therewith?

For the sake of completeness, ... I don't know all the features of your environment. If you do, however, have the chance to jump up a few compiler/Boost versions, there has really been some great progress. In particular the next and past few releases have been key since we are/have-gone standalone and C++11 on Nath/Multiprecision. So if you get a chance, some of the modern stuff might be more flexible and easier to manage.

**Comment by jzmaddock** [comment]
I can reproduce, but I'm going to close this on the basis that you've shot yourself in the foot ;)  Let me try and explain:

The overload for abs which we need to specialize looks like this:

```
template <class Backend>
typename std::enable_if<number_category<Backend>::value != number_kind_complex, number<Backend, et_off> >::type
abs(const number<Backend, et_off>& arg)
```

And if I change your unqualified call to abs with a qualified call to boost::multiprecision::abs we get a slightly clearer error message:

```
t.cpp(13): error C2672: 'boost::multiprecision::abs': no matching overloaded function found
t.cpp(13): error C2893: Failed to specialize function template 'std::enable_if<boost::multiprecision::number_category<Num>::value==,boost::multiprecision::component_type<boost::multiprecision::number<Backend,ExpressionTemplates>>>::type::type boost::multiprecision::abs(const boost::multiprecision::number<Backend,ExpressionTemplates> &)'
t.cpp(13): note: With the following template arguments:
t.cpp(13): note: 'T=boost::multiprecision::backends::cpp_bin_float<256,boost::multiprecision::backends::digit_base_2,void,int,0,0>'
t.cpp(13): note: 'ExpressionTemplates=boost::multiprecision::et_off'
```

Digging deeper, the reason it failed to specialize, is that your != operator is found for the enable_if argument where two enum's are compared.  VC2015 thinks the operator is now ambiguous, so it rejects the whole function overload.

In short, don't write overloads that are so greedy, as all kinds of unexpected interactions may occur!

I note that more recent msvc versions do not have this issue, but none the less your operators are still over-greedy IMO.

HTH.

**Comment by jzmaddock** [comment]
Here's an example which fails without boost, and which illustrates the issue:

```
template<class T, class U> static inline bool operator!=(const T& a, const U& b)
{
   return !(a == b);
};

enum number_category_type
{
   number_kind_unknown = -1,
   number_kind_integer = 0,
   number_kind_floating_point = 1,
   number_kind_rational = 2,
   number_kind_fixed_point = 3,
   number_kind_complex = 4
};

template <bool>
struct dont_do_this {};

dont_do_this<number_kind_unknown != 0> t;
```

**Comment by gpive** [comment]
Roger that.  Sorry about the blood on the carpet. : )

**Comment by ckormanyos** [comment]
Thank you John (@jzmaddock) for providing such clarity. It is good we found an explanation.
